include/dbgroup/atomic/mwcas/lock_free/mwcas_descriptor.hpp

```diff
@@ -0,0 +1,262 @@
+        (std::bit_cast<MwCASDescriptor *>(word & kAddrMask))->MwCASInternal();
+    }
+  }
+
+  /**
+   * @brief Add a new MwCAS target to this descriptor.
+   *
+   * @tparam T The class of a target word.
+   * @param addr A target memory address.
+   * @param old_val The expected value of a target field.
+   * @param new_val An inserting value into a target field.
+   * @param fence A flag for controling std::memory_order.
+   */
+  template <class T>
+  constexpr void
+  AddMwCASTarget(  //
+      void *addr,
+      const T old_val,
+      const T new_val,
+      const std::memory_order fence = std::memory_order_seq_cst)
+  {
+    static_assert(CanMwCAS<T>());
+
+    new (&(targets_.at(target_cnt_++)))
+        MwCASTarget{static_cast<std::atomic_uint64_t *>(addr), old_val, new_val, 0, fence};
+  }
+
+  /**
+   * @brief Perform a MwCAS operation by using registered targets.
+   *
+   * @retval true if a MwCAS operation succeeds.
+   * @retval false otherwise.
+   */
+  auto MwCAS()  //
+      -> bool;
+
+  /*############################################################################
+   * Internal constants
+   *##########################################################################*/
+  /// @brief A bitmask with only the "descriptor address" portion set to 1.
+  static constexpr uint64_t kAddrMask = (1UL << 47) - 1;
+
+ private:
+  /*############################################################################
+   * Internal types
+   *##########################################################################*/
+
+  /**
+   * @brief An enumeration for representing MwCAS status.
+   *
+   */
+  enum Status : uint64_t {
+    kUndecided = 0,
+    kSucceeded,
+    kFailed,
+  };
+
+  /**
+   * @brief A class for representing MwCAS targets.
+   *
+   */
+  struct MwCASTarget {
+    /// @brief A target memory address.
+    std::atomic_uint64_t *addr;
+
+    /// @brief An expected value of a target field.
+    uint64_t old_val;
+
+    /// @brief An inserting value into a target field.
+    uint64_t new_val;
+
+    /// @brief access count for GC.
+    std::atomic_uint32_t cnt;
+
+    /// @brief A fence to be inserted when embedding a new value.
+    std::memory_order fence;
+  };
+
+  /*############################################################################
+   * Internal APIs
+   *##########################################################################*/
+
+  /**
+   * @brief Embed a descriptor into this target address to linearlize MwCAS.
+   *
+   * @param desc_addr The address of this descriptor with a MwCAS flag.
+   * @param pos The position of a target word.
+   * @retval true if the descriptor address is successfully embedded.
+   * @retval false otherwise.
+   */
+  auto EmbedDescriptor(  //
+      uint64_t desc_addr,
+      size_t pos)  //
+      -> bool;
+
+  void UnembedDescriptor(   //
+      uint64_t desc_addr,   //
+      MwCASTarget &target,  //
+      int *follow_cnt,      //
+      uint64_t set_value);
```

Comment:
Doxygenコメントを追加しました．また，引数を変えていますが詳細はcppの方に書きます．

```suggestion
  /**
   * @brief Swap an embedded descriptor into a desired value.
   * 
   * @param desc_addr The address of this descriptor with the flag.
   * @param target A target MwCAS information.
   * @param desired A desired value to be embedded.
   * @return The number of followers.
   */
  auto Finalize(   //
      uint64_t desc_addr,   //
      MwCASTarget &target,  //
      uint64_t desired)  //
      -> uint32_t;
```